### ISSUE TITLE
Yolo-world results different from huggingface demo result from original repository

### ISSUE THREAD

**Issue by VasilevGrigoriy** [ultralytics/ultralytics#8663]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi!
I found that predictions from YOLOW from your library different that predictions from huggingface. I saw discussion on this theme, but my parameter 

Python code to reproduce
```python
from PIL import Image
a = [
    "anorak",
    "armband",
    "balaclava",
    "ballet",
    "bandana",
    "basque",
    "cap",
    "beret",
    "bermudas",
    "blazer",
    "blouse",
    "body",
    "bolero",
    "bomber",
    "sandals",
    "boot",
    "bracelet",
    "key ring",
    "breeches",
    "brooch",
    "trousers",
    "collar pin",
    "burkini",
    "beads",
    "boutonniere",
    "bra",
    "bustier",
    "hand fan",
    "bicycle",
    "windbreaker",
    "turtleneck",
    "tie",
    "gorget",
    "grace",
    "daflcote",
    "jeggings",
    "jumper",
    "jeans",
    "raincoat",
    "sheepskincoat",
    "jacket",
    "vest",
    "tie clip",
    "handkerchief clip",
    "cufflink",
    "badge",
    "long johns",
    "bonnet",
    "capris",
    "hood",
    "cardigan",
    "carnival costume",
    "sneaker",
    "cape",
    "kepi",
    "kigurumi",
    "kilt",
    "clutch",
    "costume jewelry button",
    "visor",
    "tights",
    "necklace",
    "ring",
    "combidress",
    "jumpsuit",
    "bathing suit",
    "underwear set",
    "corsage",
    "corset",
    "costume",
    "leather jacket",
    "sweatshirt",
    "christening dress",
    "christening shirt",
    "culottes",
    "leggings",
    "longsleeve",
    "loafers",
    "moonwalkers",
    "robe",
    "moccasin",
    "monokini",
    "mules",
    "headphone",
    "negligee",
    "petticoat",
    "nightgown",
    "stole",
    "coat",
    "panama hat",
    "pantolette",
    "folder bag",
    "pareo",
    "parka",
    "pajama",
    "piercing",
    "swimming trunks",
    "dress",
    "headband",
    "jewelry pendant",
    "poncho",
    "shoulder belt",
    "briefcase",
    "belt",
    "pullover",
    "separate swimsuit",
    "satchel",
    "rubber boot",
    "strap for suitcase",
    "shirt",
    "backpack",
    "clog",
    "sweater",
    "earring",
    "slipon",
    "one-piece swimsuit",
    "snowboot",
    "snood",
    "bag",
    "laptop bag",
    "shopping cart",
    "tankini",
    "slipper",
    "twinset",
    "sword-knot",
    "thermobody",
    "topsider",
    "top",
    "torset",
    "trench coat",
    "underpants",
    "tunic",
    "shoe",
    "ugg",
    "children-apron",
    "veil",
    "t-shirt",
    "polo t-shirt",
    "hijab",
    "hoodie",
    "chain",
    "turban",
    "watch",
    "suitcase",
    "rosary",
    "shawl",
    "hat",
    "hat-helmet",
    "charm/pendant",
    "scarve",
    "scarve-plaid",
    "flip-flop",
    "short",
    "pajama-short",
    "artificial fur coat",
    "natural fur coat",
    "cheburashka fur coat",
    "espadrille",
    "skirt",
    "glasses",
]
with torch.no_grad():
    model = YOLO(f"yolov8l-world.pt")
    model.set_classes(a)

    results = model("./image.jpg")
    results[0].save("./temp.jpg")

plt.figure(figsize=(13, 9))
plt.imshow(Image.open("./temp.jpg"))
```
Code return this predictions:
![temp](https://github.com/ultralytics/ultralytics/assets/70897555/e39bb281-d5f3-44ee-9112-7261fc185619)

This return yolov8-worldv2 with same parameters:
![temp2](https://github.com/ultralytics/ultralytics/assets/70897555/33bbc923-196e-461f-a330-a963802d0185)

But huggingface demo (and official demo in colab too) return this predictions on similar thresholds:
<img width="766" alt="Screenshot 2024-03-05 at 00 13 30" src="https://github.com/ultralytics/ultralytics/assets/70897555/ff9a2e80-db7c-4a52-8260-83f65166781e">

So, the question is why are the predictions so much different and how to fix it?

Original image to reproduce:
![Жакеты_78460355](https://github.com/ultralytics/ultralytics/assets/70897555/baf9af39-0c6d-4884-b0a0-a03c89f878ca)

**Comment by glenn-jocher** [comment]
@VasilevGrigoriy hi there! 👋

Thanks for bringing this to our attention. Differences in predictions between our library and the Hugging Face demo could stem from several factors, including model versions, preprocessing steps, or even slight differences in the inference code.

To ensure consistency, please verify the following:
1. **Model Version**: Ensure you're using the same model version across platforms. Model improvements or changes can lead to different predictions.
2. **Preprocessing**: Confirm that the image preprocessing (resizing, normalization) is identical in both cases.
3. **Thresholds**: Double-check that the confidence and IoU thresholds are the same for both your test and the Hugging Face demo.

If after checking these factors the issue persists, it might be helpful to share more details about the environment (e.g., library versions) or any modifications to the default settings.

Here's a quick check for preprocessing in your code:
```python
# Ensure consistent resizing and normalization
model.imgsz = (640, 640)  # Example size, adjust as needed
model.conf = 0.25  # Confidence threshold
```

If everything seems aligned and the issue remains, it could be beneficial to delve deeper into the model's version history or reach out for specific insights on the model's training and deployment on Hugging Face.

Hope this helps! Let us know if there's anything else we can assist with.

**Comment by VasilevGrigoriy** [comment]
Okey, I realized that use:
```python
...
model("./image.jpg")
...
```
is incorrect cause in this call I use different preprocessing (unlike original in https://github.com/AILab-CVC/YOLO-World)

To get good results, you have to do original preprocessing like:
```python
...
image = cv2.imread('/content/test.jpg')
image = cv2.resize(image, (640, 640))
image = image.astype(np.float32) / 255.0
results = model(image)
...
```

**Comment by glenn-jocher** [comment]
@VasilevGrigoriy hey there! 👋

You're absolutely right! The preprocessing steps are crucial for maintaining consistency with the model's expected input format. Your approach to manually handling the image preprocessing aligns well with ensuring that the input is correctly prepared for the model. Thanks for sharing this snippet; it's a great reminder of the importance of preprocessing steps. Keep up the great work! 🚀

If you have any more questions or need further assistance, feel free to ask. Happy coding!

**Comment by prateek-singh-inviz** [comment]
Hello @VasilevGrigoriy, Could you possibly share the preprocessing steps you’ve used? I seem to be encountering a similar problem.

**Comment by prateek-singh-inviz** [comment]
@glenn-jocher I'm facing the same issue and steps mentioned by @VasilevGrigoriy didn't helped in my case.
Versions:
Ultralytics:8.1.24
Torch:2.0.0
Code:
'''
from ultralytics import YOLO
import torch
model = YOLO('yolov8l-world.pt')  
class_names = ['shirt',
'tshirt',
'kurta',
'kurti',
'dresses',
'jacket',
'jeans',
'trousers',
'pants',
'bottomwear',
'skirts',
'leggings',
'pyjama',
'shoes',
'slippers',
'socks',
'rings',
'earrings',
'bangles',
'pendants',
'handbags',
'purse',
'hat',
'sunglasses']
model.set_classes(class_names)

image = cv2.imread('./image.jpg')
image = cv2.resize(image, (640, 640))
results = model.predict(image,conf=0.01, save=False, iou=0.1, visualize=False)

'''
Result from my code:
![image](https://github.com/ultralytics/ultralytics/assets/115148773/70df3e9f-f47b-42e6-a210-edf82ecfe280)
Result from huggingface demo:
![image](https://github.com/ultralytics/ultralytics/assets/115148773/d8727de8-3b01-4dea-8697-a25f06190806)
All the params are similar.

**Comment by VasilevGrigoriy** [comment]
Hi, @prateek-singh-inviz! 

You can try to add background class like this - https://github.com/ultralytics/ultralytics/issues/8651.

Also, you can add argument agnostic_nms=True in model predict to use agnostic non-maximum-suppression, because here there are similar classes (for example pants and leggings), so if model predict these classes on one image, then common nms will not perform postprocessing on them and will leave both preds (because common nms only works for objects of the same class)

**Comment by prateek-singh-inviz** [comment]
@VasilevGrigoriy adding a background class along with agnostic_nms solved the issue. Thank you

**Comment by youyuge34** [comment]
> Hi, @prateek-singh-inviz!
> 
> You can try to add background class like this - #8651.
> 
> Also, you can add argument agnostic_nms=True in model predict to use agnostic non-maximum-suppression, because here there are similar classes (for example pants and leggings), so if model predict these classes on one image, then common nms will not perform postprocessing on them and will leave both preds (because common nms only works for objects of the same class)

I got the same problem too and add background class is the final solve solution!

**Comment by eliabntt** [comment]
> @VasilevGrigoriy hey there! 👋
> 
> You're absolutely right! The preprocessing steps are crucial for maintaining consistency with the model's expected input format. Your approach to manually handling the image preprocessing aligns well with ensuring that the input is correctly prepared for the model. Thanks for sharing this snippet; it's a great reminder of the importance of preprocessing steps. Keep up the great work! 🚀
> 
> If you have any more questions or need further assistance, feel free to ask. Happy coding!

I think it might make sense to add this to the [Yolo-World](https://docs.ultralytics.com/models/yolo-world/#available-models-supported-tasks-and-operating-modes) guide, perhaps including the correspondence btw the model you listed and the one on the original repo (including imgsz factors)
